src/atlas/array/helpers/ArrayForEach.h

```diff
@@ -0,0 +1,276 @@
+/*
+ * (C) Crown Copyright 2023 Met Office
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+#pragma once
+
+#include <array>
+#include <tuple>
+#include <type_traits>
+
+#include "atlas/array/ArrayView.h"
+#include "atlas/array/Range.h"
+#include "atlas/array/helpers/ArraySlicer.h"
+#include "atlas/parallel/omp/omp.h"
+#include "atlas/util/Config.h"
+
+#include "eckit/exception/Exceptions.h"
+
+namespace atlas {
+namespace execution {
+
+class sequenced_policy {};
+class unsequenced_policy {};
+class parallel_unsequenced_policy {};
+class parallel_policy {};
+
+// execution policy objects
+inline constexpr sequenced_policy            seq{ /*unspecified*/ };
+inline constexpr parallel_policy             par{ /*unspecified*/ };
+inline constexpr parallel_unsequenced_policy par_unseq{ /*unspecified*/ };
+inline constexpr unsequenced_policy          unseq{ /*unspecified*/ };
+
+
+template <typename execution_policy>
+std::string policy_name() {
+  return "unsupported";
+};
+template <>
+std::string policy_name<sequenced_policy>() {
+  return "sequenced_policy";
+};
+template <>
+std::string policy_name<unsequenced_policy>() {
+  return "unsequenced_policy";
+};
+template <>
+std::string policy_name<parallel_unsequenced_policy>() {
+  return "parallel_unsequenced_policy";
+};
+
+template <typename execution_policy>
+std::string policy_name(execution_policy) {
+  return policy_name<execution_policy>();
+}
+
```

Comment:
@wdeconinck I like what you did with the policies. I'd left them a bit of a mess!